codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ProtocolDocumentGenerator.java

```diff
@@ -0,0 +1,520 @@
+     * <p>
+     * This package is not generated if the service does not have any document shapes in the model.
+     *
+     * <pre>{@code
+     * // <servicePath>/document
+     * package document
+     *
+     * import (
+     *      internaldocument "<servicePath>/internal/document"
+     * )
+     *
+     * type Interface = internaldocument.Interface
+     *
+     * func NewLazyDocument(v interface{}) Interface {
+     *      return internaldocument.NewDocumentMarshaler(v)
+     * }
+     * }</pre>
+     */
+    private void generateDocumentPackage() {
+        if (!this.hasDocumentShapes) {
+            return;
+        }
+
+        writeDocumentPackage("doc.go", writer -> {
+            String documentTemplate = IoUtils.readUtf8Resource(getClass(), "document_doc.go.template");
+            writer.writeRawPackageDocs(documentTemplate);
+        });
+
+        writeDocumentPackage("document.go", writer -> {
+            writer.writeDocs(String.format("%s defines a document which is a protocol-agnostic type which supports a "
+                    + "JSON-like data-model. You can use this type to send UTF-8 strings, arbitrary precision "
+                    + "numbers, booleans, nulls, a list of these values, and a map of UTF-8 strings to these "
+                    + "values.", DOCUMENT_INTERFACE_NAME));
+            writer.writeDocs("");
+            writer.writeDocs(String.format("You create a document type using the %s function and passing it the Go "
+                    + "type to marshal. When receiving a document in an API response, you use the "
+                    + "document's UnmarshalSmithyDocument function to decode the response to your desired Go "
+                    + "type. Unless documented specifically generated structure types in client packages or "
+                    + "client types packages are not supported at this time. Such types embed a "
+                    + "noSmithyDocumentSerde and will cause an error to be returned when attempting to send an "
+                    + "API request.", NEW_LAZY_DOCUMENT));
+            writer.writeDocs("");
+            writer.writeDocs("For more information see the accompanying package documentation and linked references.");
+            writer.write("type $L = $T", DOCUMENT_INTERFACE_NAME,
+                            getInternalDocumentSymbol(DOCUMENT_INTERFACE_NAME))
+                    .write("");
+
+            writer.writeDocs(String.format("You create document type using the %s function and passing it the Go "
+                    + "type to be marshaled and sent to the service. The document marshaler supports semantics similar "
+                    + "to the encoding/json Go standard library.", NEW_LAZY_DOCUMENT));
+            writer.writeDocs("");
+            writer.writeDocs("For more information see the accompanying package documentation and linked references.");
+            writer.openBlock("func $L(v interface{}) $T {", "}", NEW_LAZY_DOCUMENT,
```

Comment:
Should this document the struct tag to use for custom member names?

Reply:
The doc.go for the clients `document` package has a `doc.go` from this template file https://github.com/aws/smithy-go/blob/a87ef15659d457c607421ee90635d35b2e6def4c/codegen/smithy-go-codegen/src/main/resources/software/amazon/smithy/go/codegen/document_doc.go.template which details more details and links to the smithy-go document package.